components/raftstore/src/store/txn_ext.rs

```diff
@@ -188,6 +190,16 @@ impl PeerPessimisticLocks {
             let desc = key.len() + lock.memory_size();
             self.memory_size -= desc;
             GLOBAL_MEM_SIZE.sub(desc as i64);
+
+            // Shrink HashMap if it's significantly under-utilized to prevent
+            // memory leaks
+            // Only shrink if capacity is reasonably large to avoid thrashing
+            // on small maps
+            let len = self.map.len();
+            let capacity = self.map.capacity();
+            if capacity >= MAP_SHRINK_THRESHOLD && len * 4 < capacity {
```

Comment:
How about only trying to shrink when the capacity or allocated space is close to the `PER_REGION_PESSIMISTIC_SIZE_LIMIT` ?

Reply:
The current shrink seems to be easy to get triggerred, my concern is that shrinking frequently may introduce performance degradation leading to performance oncall issues.

IMO The trade-off strategy is to control the approximate memory usage under `size limit` to avoid the worst case together with the `STORE_PESSIMISTIC_SIZE_LIMIT` configuration.

Reply:
Let me benchmark it